ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/consumer/ScalablePushConsumer.java

```diff
@@ -121,47 +132,89 @@ public void run() {
       LOG.error("Already ran consumer");
       throw new IllegalStateException("Already ran consumer");
     }
-    started = true;
-    initialize();
-    subscribeOrAssign();
-    while (!closed) {
-      final ConsumerRecords<?, GenericRow> records = consumer.poll(POLL_TIMEOUT);
-      // No assignment yet
-      if (this.topicPartitions.get() == null) {
-        continue;
-      }
-      if (records.isEmpty()) {
+    try {
+      started = true;
+      initialize();
+      subscribeOrAssign();
+      while (!closed) {
+        final ConsumerRecords<?, GenericRow> records = consumer.poll(POLL_TIMEOUT);
+        // No assignment yet
+        if (this.topicPartitions.get() == null) {
+          continue;
+        }
+
+        if (newAssignment) {
+          newAssignment = false;
+          onNewAssignment();
+        }
+
+        final PushOffsetVector startOffsetVector
+            = getOffsetVector(currentPositions, topicName, partitions);
+        if (records.isEmpty()) {
+          updateCurrentPositions();
+          computeProgressToken(Optional.of(startOffsetVector));
+          onEmptyRecords();
+          continue;
+        }
+
+        for (ConsumerRecord<?, GenericRow> rec : records) {
+          handleRow(rec.key(), rec.value(), rec.timestamp());
+        }
+
         updateCurrentPositions();
-        onEmptyRecords();
-        continue;
+        computeProgressToken(Optional.of(startOffsetVector));
+        try {
+          consumer.commitSync();
+          updateCommittedOffsets();
+        } catch (CommitFailedException e) {
+          LOG.warn("Failed to commit, likely due to rebalance.  Will wait for new assignment", e);
+        }
+
+        afterCommit();
```

Comment:
should we call after commit even if the commit failed? if so maybe rename it?

Reply:
Changed it to `afterBatchProcessed`